File: Notes/CTCI/03-stacks-queues/stack.js
module.exports = function Stack() {
  const items = [];

  // Push an item to the top of the stack
  this.push = function(data) {
    items.push(data);
  }

  // Return item at the top of the stack
  this.pop = function() {
    if (this.isEmpty()) return null;

    return items.pop();
  }

  // Look at item at the top of the stack
  this.peek = function() {
    return items[items.length - 1];
  }

  this.isEmpty = function() {
    return items.length === 0;
  }
}

File: Notes/CTCI/03-stacks-queues/py/Stack.py
class Stack:
    def __init__(self):
        self.items = []

    def push(self, i):
        return self.items.add(i)

    def pop(self):
        return None if self.is_empty() else self.items.pop()

    def is_empty(self):
        return len(self.items) == 0

    def peek(self):
        return self.items[-1]
